### DISCUSSION TITLE
mysql: P1000: Authentication failed

### DISCUSSION THREAD

**Discussion by jelmd** [prisma/prisma#21961] [Q&A]
### Bug description

Prisma seems to be unable to connect to remote databases. Neither migration nor studio, etc. work and always throw a `P1000: Authentication failed` error. It is a prisma bug because connecting with the native mysql client and the same parameters as set for prisma is no problem at all, i.e. native mysql client works as expected.

### How to reproduce

npx prisma migrate dev

### Expected behavior

migration should simply work.

### Prisma information

```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "mysql"
  url      = env("DATABASE_URL")
  shadowDatabaseUrl = env("DATABASE_TMP_URL")
}

model Course {
  id           Int    @id @default(autoincrement())
  cName        String
}
```

### Environment & setup

- OS: Ubuntu 20.04 (focal)
- Database: MySQL 8.0.34-0ubuntu0.20.04.1
- Node.js version: v18.17.1

.env:
```sh
DATABASE_URL=mysql://user:password@dbhost:3306/dbname?connection_limit=10
DATABASE_TMP_URL=mysql://user:password@dbhost:3306/dbname_tmp?connection_limit=10
```
Replace user, password, dbhost, dbname accordingly.

### Prisma Version

```
Environment variables loaded from .env
prisma                  : 5.6.0
@prisma/client          : 5.6.0
Computed binaryTarget   : debian-openssl-1.1.x
Operating System        : linux
Architecture            : x64
Node.js                 : v18.17.1
Query Engine (Node-API) : libquery-engine xxx (at node_modules/@prisma/engines/libquery_engine-debian-openssl-1.1.x.so.node)
Schema Engine           : schema-engine-cli xxx (at node_modules/@prisma/engines/schema-engine-debian-openssl-1.1.x)
Schema Wasm             : @prisma/prisma-schema-wasm 5.6.0-32.xxx
Default Engines Hash    : xxx
Studio                  : 0.495.0

```

**Comment by jelmd** [comment]
Why got it moved to discussion? Isn't it a bug, if it doesn't work?

**Reply by ludralph** [reply]
Hi @jelmd 

This was moved to a discussion because connecting to remote databases work and is not broken.

**Reply by janpio** [reply]
Here is a recent test that shows Prisma Client talking to a remote MySQL database at PlanetScale: https://github.com/prisma/ecosystem-tests/actions/runs/6876133711/job/18701197383

If you have trouble with talking to your MySQL database, it is probably a problem that needs to be understood and debugged first before we consider this a bug - hence I turned this into a discussion.

**Reply by jelmd** [reply]
Hmmm, can't find the setup of the mysql db there. So no clue about version or configuration. From `GCP_MYSQL_SSL_DB_URL` I guess, an instance hosted and mangled by Google is used? Hmmm, if so I wouldn't trust it. IMHO it would be better to setup the db within the test environment using the packages provided by the related Distro vendor and test against this one.

**Reply by janpio** [reply]
That is exactly what the tests in this repository do: https://github.com/prisma/prisma/blob/d4f2f3be1704286292bb727f0e4f38930e1c4fc6/docker/docker-compose.yml#L96-L111

Believe me, we would be aware if Prisma generally could not talk to MySQL databases.

**Comment by ludralph** [comment]
Hi @jelmd 👋 

Thank you for raising this question.

> Prisma seems to be unable to connect to remote databases.

You mentioned that Prisma seems to be unable to connect to remote databases. There are some limitations that apply for some cloud hosted databases. Can you please take a look at this section of the [documentation](https://www.prisma.io/docs/concepts/components/prisma-migrate/shadow-database#cloud-hosted-shadow-databases-must-be-created-manually) for using shadow database with a cloud hosted service.

Also, the database user must have CREATE, ALTER, DROP, REFERENCES ON *.* privilege for MYSQL as shown in this section of the [documentation](https://www.prisma.io/docs/concepts/components/prisma-migrate/shadow-database#shadow-database-user-permissions).

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Comment by jelmd** [comment]
> Hi @jelmd 👋
> 
> Thank you for raising this question.
> 
> > Prisma seems to be unable to connect to remote databases.
> 
> You mentioned that Prisma seems to be unable to connect to remote databases. There are some limitations that apply for some cloud hosted databases. Can you please take a look at this section of the [documentation](https://www.prisma.io/docs/concepts/components/prisma-migrate/shadow-database#cloud-hosted-shadow-databases-must-be-created-manually) for using shadow database with a cloud hosted service.

As you can certainly see in the problem description, that a shadow db has been setup (see schema and .env file). Furthermore if this would be the problem, wouldn't this throw a different error than P1000? As shown, it is not a connection but an authentication problem - at least the error message says this and monitoring the traffic between prisma client and db host clearly shows, that a connection gets established and pakets sent in both directions.

BTW: Not sure, what your definition of "cloud based" is. Our db server is a normal box in our local network running Ubuntu 20.04. It has several DB instances running used by several remote service (web servers, wordpress, slurm, etc.) in production without any problems.

> Also, the database user must have CREATE, ALTER, DROP, REFERENCES ON _._ privilege for MYSQL as shown in this section of the [documentation](https://www.prisma.io/docs/concepts/components/prisma-migrate/shadow-database#shadow-database-user-permissions).

Well, if this would be the problem, it would throw a different error, but not a P1000 error, right? Or is prisma trying to modify the db before logged in to it?

> If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved.

No, IMHO it is a bug. As said, with `mysql` client (even with `--ssl-mode=DISABLED`) one is able to connect to the normal as well as to the shadow database w/o any problems. So the DBs are there, the username and password are correct (actually they are the same for both DBs). 

> Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Comment by janpio** [comment]
Can you please share the full error you are getting?

**Comment by jelmd** [comment]
Yepp:
```
app # npm run migrate:dev
> app@0.1.0 migrate:dev
> prisma migrate dev

Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma
Datasource "db": MySQL database "dbtest" at "dbhost:3306"

Error: P1000: Authentication failed against database server at `dbhost`, the provided database credentials for `dbuser` are not valid.

Please make sure to provide valid database credentials for the database server at `dbhost`.
```

**Reply by janpio** [reply]
Do you get the same error when you run `db pull`, or is this specific to `migrate dev`?
If you get the same error, then we can conclude that the shadow database is not relevant here at all.)

**Reply by hozaifa4you** [reply]
Yes

```sh
$ npx prisma db pull
Prisma schema loaded from prisma/schema.prisma
Environment variables loaded from .env
Datasource "db": MySQL database "aipdfquery_staging" at "localhost:3306"

✖ Introspecting based on datasource defined in prisma/schema.prisma

Error: P1000

Authentication failed against database server at `localhost`, the provided database credentials for `aipdfquery_staging` are not valid.

Please make sure to provide valid database credentials for the database server at `localhost`.

$ npx prisma db push
Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma
Datasource "db": MySQL database "aipdfquery_staging" at "localhost:3306"

Error: P1000: Authentication failed against database server at `localhost`, the provided database credentials for `aipdfquery_staging` are not valid.

Please make sure to provide valid database credentials for the database server at `localhost`.
```

**Comment by jelmd** [comment]
Yes same beside the `✖ Introspecting based on datasource defined in prisma/schema.prisma` line.

**Comment by jelmd** [comment]
Same thing with studio. I.e. as soon as you click on something which needs the db connection, the same error appears.

**Reply by janpio** [reply]
(Could you please start responding in thread instead of opening a new one with each message? I don't get notifications about responses this way. Thank you.)

Ok, then we know that Prisma can not talk to your main `url` database.

You shared this placeholder string above: `mysql://user:password@dbhost:3306/dbname`
Any special characters in any of the parts that you replace here?

**Comment by jelmd** [comment]
> (Could you please start responding in thread instead of opening a new one with each message? I don't get notifications about responses this way. Thank you.)

Ehmm, you mean I should use 'Quote reply' instead of 'Write a reply'? Hmmm, never noticed a difference wrt. notifications. However, the original post gets sometimes copied twice into the 'suggested answer' area ... =8-( .

> Ok, then we know that Prisma can not talk to your main `url` database.
> 
> You shared this placeholder string above: `mysql://user:password@dbhost:3306/dbname` Any special characters in any of the parts that you replace here?

Not really, all ASCII in the range of 0x21..0x7f excl. any quote chars (user,dbname,dbhost in the range of [_a-z]).

**Reply by janpio** [reply]
> Ehmm, you mean I should use 'Quote reply' instead of 'Write a reply'? Hmmm, never noticed a difference wrt. notifications. However, the original post gets sometimes copied twice into the 'suggested answer' area ... =8-( .

No, there is a "Write a reply" input field under messages here as this is a GitHub discussion, not a GitHub issue where there is only on stream of comments.

**Comment by jelmd** [comment]
> That is exactly what the tests in this repository do:
> 
> https://github.com/prisma/prisma/blob/d4f2f3be1704286292bb727f0e4f38930e1c4fc6/docker/docker-compose.yml#L96-L111
> 
> Believe me, we would be aware if Prisma generally could not talk to MySQL databases.

Well, didn't say that, but that there is a bug wrt. authentication. And looking at the line 98 - `command: --default-authentication-plugin=mysql_native_password` I think this is the problem. The [MySQL Reference Manual 8.0](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_default_authentication_plugin) says 

> In MySQL 8.0, caching_sha2_password is the default authentication plugin rather than mysql_native_password.

So I guess prisma tries blindly the native password instead of the sha2 password.

**Reply by janpio** [reply]
That's just legacy, see this PR for it working without as well: https://github.com/prisma/prisma/pull/21975

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
